core/src/main/java/org/elasticsearch/cluster/routing/UnassignedMeta.java

```diff
@@ -0,0 +1,167 @@
+         */
+        DANGLING_IMPORT,
+        /**
+         * Unassigned as a result of restoring into a new index.
+         */
+        NEW_RESTORE,
+        /**
+         * Unassigned as a result of restoring into a closed index.
+         */
+        EXISTING_RESTORE,
+        /**
+         * Unassigned as a result of explicit addition of a replica.
+         */
+        ADD_REPLICA,
+        /**
+         * Unassigned as a result of a failed allocation of the shard.
+         */
+        FAILED,
+        /**
+         * Unassigned as a result of the node hosting it leaving the cluster.
+         */
+        NODE_LEAVE,
+        /**
+         * Unassigned as a result of explicit cancel reroute command.
+         */
+        EXPLICIT_CANCEL;
+    }
+
+    private final Reason reason;
+    private final long timestamp;
+    private final String details;
+
+    public UnassignedMeta(Reason reason, String details) {
+        this(reason, System.currentTimeMillis(), details);
+    }
+
+    private UnassignedMeta(Reason reason, long timestamp, String details) {
+        this.reason = reason;
+        this.timestamp = timestamp;
+        this.details = details;
+    }
+
+    UnassignedMeta(StreamInput in) throws IOException {
+        this.reason = Reason.values()[(int) in.readByte()];
+        this.timestamp = in.readLong();
+        this.details = in.readOptionalString();
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeByte((byte) reason.ordinal());
+        out.writeLong(timestamp);
+        out.writeOptionalString(details);
+    }
+
+    public UnassignedMeta readFrom(StreamInput in) throws IOException {
+        return new UnassignedMeta(in);
+    }
+
+    /**
+     * The reason why the shard is unassigned.
+     */
+    public Reason getReason() {
+        return this.reason;
+    }
+
+    /**
+     * The timestamp in milliseconds since epoch. Note, we use timestamp here since
+     * we want to make sure its preserved across node serializations. Extra care need
+     * to be made if its used to calculate diff (handle negative values) in case of
+     * time drift.
+     */
+    public long getTimestampInMillis() {
+        return this.timestamp;
+    }
+
+    /**
+     * Returns optional details explaining the reasons.
+     */
+    @Nullable
+    public String getDetails() {
+        return this.details;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("unassigned[[reason=").append(reason).append("]");
+        sb.append(", at[").append(DATE_TIME_FORMATTER.printer().print(timestamp)).append("]");
+        if (details != null) {
+            sb.append(", details[").append(details).append("]");
+        }
+        sb.append("]");
+        return sb.toString();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("unassigned");
```

Comment:
I think the unassigned name will be confusing for an intializing shard. Can't find a better name though.. pointing it out in the hope that someone else can. If not, I'm good with this.